contracts/extensions/Auctions/DutchAuction.sol

```diff
@@ -0,0 +1,483 @@
+   * @param _memo Text to publish as part of the creation event.
+   */
+  function create(
+    IERC721 collection,
+    uint256 item,
+    uint256 startingPrice,
+    uint256 endingPrice,
+    uint256 _duration,
+    JBSplit[] calldata saleSplits,
+    string calldata _memo
+  ) external override nonReentrant {
+    if (!getBoolean(settings, 32) && !hasRole(AUTHORIZED_SELLER_ROLE, msg.sender)) {
+      revert NOT_AUTHORIZED();
+    }
+
+    bytes32 auctionId = keccak256(abi.encodePacked(address(collection), item));
+    DutchAuctionData memory auctionDetails = auctions[auctionId];
+
+    if (auctionDetails.info != 0) {
+      revert AUCTION_EXISTS();
+    }
+
+    if (startingPrice > type(uint96).max) {
+      revert INVALID_PRICE();
+    }
+
+    if (endingPrice > type(uint96).max || endingPrice >= startingPrice) {
+      revert INVALID_PRICE();
+    }
+
+    // TODO: Check that expiration - now > periodDuration
+
+    {
+      // scope to reduce stack depth
+      uint256 auctionInfo = uint256(uint160(msg.sender));
+      auctionInfo |= uint256(uint64(block.timestamp - deploymentOffset)) << 160;
+
+      uint256 auctionPrices = uint256(uint96(startingPrice));
+      auctionPrices |= uint256(uint96(endingPrice)) << 96;
+      auctionPrices |= uint256(uint64(block.timestamp - deploymentOffset + _duration)) << 192;
+
+      auctions[auctionId] = DutchAuctionData(auctionInfo, auctionPrices, 0);
+    }
+
+    uint256 length = saleSplits.length;
+    for (uint256 i = 0; i < length; i += 1) {
+      auctionSplits[auctionId].push(saleSplits[i]);
+    }
+
+    collection.transferFrom(msg.sender, address(this), item);
+
+    emit CreateDutchAuction(
+      msg.sender,
+      collection,
+      item,
+      startingPrice,
+      endingPrice,
+      block.timestamp + _duration,
+      _memo
+    );
+  }
+
+  /**
+   * @notice Places a bid on an existing auction. Refunds previous bid if needed. The contract will only store the highest bid. The bid can be below current price in anticipation of the auction eventually reaching that price. The bid must be at or above the end price.
+   *
+   * @param collection ERC721 contract.
+   * @param item Token id to bid on.
+   */
+  function bid(
+    IERC721 collection,
+    uint256 item,
+    string calldata _memo
+  ) external payable override nonReentrant {
+    bytes32 auctionId = keccak256(abi.encodePacked(collection, item));
+    DutchAuctionData memory auctionDetails = auctions[auctionId];
+
+    if (auctionDetails.info == 0) {
+      revert INVALID_AUCTION();
+    }
+
+    uint256 expiration = uint256(uint64(auctionDetails.prices >> 192));
+
+    if (block.timestamp > deploymentOffset + expiration) {
+      revert AUCTION_ENDED();
+    }
+
+    if (auctionDetails.bid != 0) {
+      uint256 currentBidAmount = uint96(auctionDetails.bid >> 160);
+      if (currentBidAmount >= msg.value) {
+        revert INVALID_BID();
+      }
+
+      payable(address(uint160(auctionDetails.bid))).transfer(currentBidAmount);
```

Comment:
should i replace it with a `call`; best effort but ignore failure?

Reply:
If you do a call, make sure to limit the gas it can use, otherwise it might introduce reentrence. Alternatively you can also keep a mapping with userRefunds and users then have to manually call a method to get their refund. Although not very UX friendly it does make it easier to safely handle refunds.